crates/bitwarden-ipc/src/traits/communication_backend.rs

```diff
@@ -65,19 +111,33 @@ pub mod tests {
 
     impl CommunicationBackend for TestCommunicationBackend {
         type SendError = ();
-        type ReceiveError = TestCommunicationBackendReceiveError;
+        type Receiver = TestCommunicationBackendReceiver;
 
         async fn send(&self, message: OutgoingMessage) -> Result<(), Self::SendError> {
             self.outgoing.write().await.push(message);
             Ok(())
         }
 
+        async fn subscribe(&self) -> Self::Receiver {
+            TestCommunicationBackendReceiver(RwLock::new(self.incoming_rx.resubscribe()))
+        }
+    }
+
+    impl CommunicationBackendReceiver for TestCommunicationBackendReceiver {
+        type ReceiveError = TestCommunicationBackendReceiveError;
+
         async fn receive(&self) -> Result<IncomingMessage, Self::ReceiveError> {
-            if let Some(message) = self.incoming.write().await.pop_front() {
-                Ok(message)
-            } else {
-                Err(TestCommunicationBackendReceiveError::NoQueuedMessages)
+            if self.0.read().await.len() == 0 {
+                return Err(TestCommunicationBackendReceiveError::NoQueuedMessages);
             }
+
+            Ok(self
+                .0
+                .write()
+                .await
+                .recv()
+                .await
+                .expect("Failed to receive incoming message"))
```

Comment:
We should grab the rwlock only once to avoid issues with other requests getting in between. Also use `.is_empty()` instead of `.len() == 0`

```suggestion
            let mut guard = self.0.write().await;

            if guard.is_empty() {
                return Err(TestCommunicationBackendReceiveError::NoQueuedMessages);
            }

            Ok(guard
                .recv()
                .await
                .expect("Failed to receive incoming message"))
        }
```